engines/kingdom/GPLogic4.cpp

```diff
@@ -0,0 +1,376 @@
+		_vm->drawPic(179);
+
+	_vm->_iconRedraw = true;
+	if (_vm->_userInput == 0x2F1) {
+		_vm->_asMode = false;
+		_vm->restoreAS();
+		_statPlay = _oldStatPlay;
+		_vm->_loopFlag = true;
+	} else {
+		if (_health == 10 || _health == 6 || _health == 2 || _vm->_userInput < 0x400 || _vm->_userInput > 0x427) {
+			_vm->_asMode = false;
+			_vm->restoreAS();
+			_statPlay = _oldStatPlay;
+		} else {
+			_statPlay = _mapExit[_vm->_userInput - 0x400];
+			_vm->_userInput = 0;
+			wound();
+			_vm->playMovie(10);
+			inventoryDel(3);
+			_vm->_tsIconOnly = false;
+			enAll();
+			switchAS();
+			_mapStat = 0;
+		}
+		_vm->_userInput = 0;
+		_vm->_loopFlag = true;
+	}
+}
+
+void Logic::GPL4_991() {
+	_currMap = 10;
+	_vm->_userInput = 0;
+	_vm->playSound(0);
+	if (_health == 10 || _health == 6 || _health == 2)
+		_health -= 2;
+	else
+		_health -= 4;
+
+	if (_health == 0) {
+		_vm->playMovie(194);
+		_vm->_noIFScreen = true;
+		_vm->fShowPic(107);
+		_currMap = 3;
+		dsAll();
+		_pouch = false;
+		_vm->playSound(1);
+		memset(_inventory, -1, 19);
+		_statPlay = 994;
+	} else {
+		_vm->_bTimer = 36;
+		while(_vm->_bTimer) {
+			_vm->checkTimers();
+			_vm->refreshSound();
+		}
+
+		_vm->_loopFlag = true;
+
+		if (_nodes[28] == 1) {
+			_nodes[28] = 0;
+			_nodes[48] = 0;
+		}
+
+		_vm->_itemInhibit = false;
+		_tideCntl = false;
+		_vm->drawPic(179);
+		_vm->_skylarTimer = 0;
+		_vm->_skylarTimerFlag = false;
+		enAll();
+		_vm->_mapEx = true;
+		_nodeNum = 5;
+		_vm->drawLocation();
+		_vm->playSound(0);
+		if (_nodes[5] > 0) {
+			_vm->playMovie(197);
+			_resurrect = true;
+		}
+		inventoryDel(12);
+		_statPlay = 50;
+	}
+}
+
+void Logic::GPL4_992() {
+	_currMap = 10;
+	_vm->_userInput = 0;
+	_vm->playSound(0);
+	if (_health == 10 || _health == 6 || _health == 2)
+		_health -= 2;
+	else
+		_health -= 4;
+
+	if (_health == 0) {
+		_vm->playMovie(194);
+		_vm->_noIFScreen = true;
+		_vm->fShowPic(107);
+		_currMap = 3;
+		dsAll();
+		_pouch = false;
+		_vm->playSound(1);
+		memset(_inventory, -1, 19);
+		_statPlay = 994;
+	} else {
+		_vm->_bTimer = 36;
+		while(_vm->_bTimer != 0) {
+			_vm->checkTimers();
+			_vm->refreshSound();
+		}
+
+		_vm->_loopFlag = true;
+		if (_nodes[28] == 1) {
+			_nodes[28] = 0;
+			_nodes[48] = 0;
+		}
+
+		_vm->_itemInhibit = false;
+		_tideCntl = false;
+		_vm->drawPic(179);
+		_vm->_skylarTimer = 0;
+		_vm->_skylarTimerFlag = false;
+		enAll();
+		_vm->_mapEx = true;
+		_nodeNum = 27;
+		_vm->drawLocation();
+		_resurrect = true;
+		_vm->playSound(0);
+		_vm->playMovie(195);
+		inventoryDel(12);
+		_statPlay = 270;
+	}
+}
+
+void Logic::GPL4_992_demo() {
+	_currMap = 10;
+	_vm->_userInput = 0;
+	_vm->playSound(0);
+	if (_health == 10 || _health == 6 || _health == 2)
+		_health -= 2;
+	else
+		_health -= 4;
+
+	if (_health == 0) {
+		_vm->playMovie(194);
+		_vm->_noIFScreen = true;
+		_vm->fShowPic(107);
+		_currMap = 3;
+		dsAll();
+		_pouch = false;
+		_vm->playSound(1);
+		memset(_inventory, -1, 19);
+		_statPlay = 994;
```

Comment:
This block handles character death, and is repeated throughout the game logic. It would be better to put it into its own function, e.g. playerDied().

Also, we infer that movie 194 and pic 107 are the "Player death" ones, and that screen 994 is the starting one, so it's easy to map these to an enum as well.

Reply:
I won't add the enum right now because I'm not able to play the video and the picture isn't self-explaining. But the code has been moved.